README.md

```diff
@@ -14,18 +14,55 @@ Implement the tool `st` that does exact pattern matching using a suffix tree. Te
 # Report
 
 ## Specify if you have used a linear time or quadratic time algorithm.
+We ended up with a quadratic time algorithm, we played around with mccreight, but other projects got the better of us, and ate what little time we had left.
 
 ## Insights you may have had while implementing and comparing the algorithms.
+Our implementation choice on how to handle leaves (in this case they're just ints) sparked a lengthy yet very enlightening discussion on how we interpret and envision abstract data structure in our minds eye, turns out we each have our differences.  
```

Comment:
If the leaves are just integers, I would also be a bit confused. You need two pieces of information in a leaf: the suffix index and the start position of the string going down the edge to that leaf. The suffix index alone does not give you the second, so you would need that from somewhere else. Where do you get it?

**Edit:** Ah, now I see how it works. You are lying to me here. It's not that leaves are represented as integers, which would be strange, but that a leaf's `children` is an integer. 

Yeah, that is a bit strange, but only because we are in Python. Many languages have a *union* type where we can use the same memory to represent two different things. They do allow us to give the different data different names, though.

You can as well, if you want. You can use a property.

```python
class Foo:
     _union: int | dict

    @property
    def children(self) -> dict:
        assert type(self._union) == dict
        return cast(dict, self._union)

    @property
    def label(self) -> int:
        assert type(self._union) == int
        return cast(int, self._union)
```

Not completely fleshed out, but this is the idea. You would use `foo.children` to get the dict and `foo.label` to get the integer, but there is only ever one of the two.